### DISCUSSION TITLE
SQLAlchemy: `db.refresh()` necessary after `db.commit()`?

### DISCUSSION THREAD

**Discussion by briantjt** [fastapi/fastapi#8905] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

According to [FastAPI's SQLAlchemy tutorial](https://fastapi.tiangolo.com/tutorial/sql-databases/#create-data), `db.refresh()` is called after `db.commit()`. However from [SQLAlchemy's documentation](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#committing):

> Finally, all objects within the Session are expired as the transaction is closed out. This is so that when the instances are next accessed, either through attribute access or by them being present in the result of a SELECT, they receive the most recent state. This behavior may be controlled by the Session.expire_on_commit flag, which may be set to False when this behavior is undesirable. 

So in this case, `db.refresh()` shouldn't be necessary as `expire_on_commit` defaults to `True` and the tutorial doesn't set this to `False`? Or does FastAPI/Pydantic do something to the ORM object that necessitates this?

**Comment by tusharsnx** [comment]
@brianjt
until there is no `db.commit()` the data that is fetched once remains in the memory and never accessed from the `database` but once `db.commit()` is executed, first access request of every data(in `ORM` it's an object) will be fetched from the `database` and remains in the memory until next `db.commit()`. Every subsequent query will fetch it from the memory instead of making another object for the same data. I never had to do `db.refresh()` after `db.commit()` but it could be useful in some cases.
